libraries/AP_NavEKF3/AP_NavEKF3_core.cpp

```diff
@@ -1800,9 +1801,31 @@ void NavEKF3_core::ForceSymmetry()
 void NavEKF3_core::ConstrainVariances()
 {
     for (uint8_t i=0; i<=3; i++) P[i][i] = constrain_float(P[i][i],0.0f,1.0f); // attitude error
-    for (uint8_t i=4; i<=6; i++) P[i][i] = constrain_float(P[i][i],0.0f,1.0e3f); // velocities
-    for (uint8_t i=7; i<=8; i++) P[i][i] = constrain_float(P[i][i],0.0f,1.0e6f);
-    P[9][9] = constrain_float(P[9][9],0.0f,1.0e6f); // vertical position
+    for (uint8_t i=4; i<=5; i++) P[i][i] = constrain_float(P[i][i], VEL_STATE_MIN_VARIANCE, 1.0e3f); // NE velocity
+
+    // check for collapse of the vertical velocity variance
+    if (P[6][6] < VEL_STATE_MIN_VARIANCE) {
+        P[6][6] = VEL_STATE_MIN_VARIANCE;
+        vertVelVarClipCount++;
```

Comment:
never decreasing this count means the chances of the reset triggering goes up with flight time. Maybe we should make it a uint32_t, and increase by 10000 each time it hits the limit, and each time it has not hit the limit we decrease by 1. That means at 400Hz we decay a single event to zero over 25s. Then we make the reset threshold be 10000x higher. That would lower the likelyhood of us inappropriately triggering over very long flight times

Reply:
EKF_TARGET_DT is currently 0.012 so I'm thinking we could increment by 1/EKF_TARGET_DT each time it hits the limit, decrement by 1 each covariance prediction cycle and set the threshold to 5/EKF_TARGET_DT so that each event is decayed over 1 second which is enough to capture the slowest sensor update that is causing a collapse.